### PR TITLE
[HOLD] feat(kda): replay-based lazy commit for speculative target verify

### PR DESCRIPTION
**Pull request by nperrin-fr**
## Summary

- Memory: ~8.8 GB/rank freed at bs=32 (T=4, 69 KDA layers). The per-position verify scratch — 795 KiB × (T+1) rows × layers per request — is replaced by a 9.3 KiB/token/layer input payload (~82 MB total, 86× smaller per token).
- Latency: neutral end-to-end; verify kernel itself ~18% faster than the scratch path (220→181 µs/layer at bs=64 T=2) since lazy fusion cuts state traffic to 1 read + 1 write per round.
- Quality unchanged: GPQA-diamond 92.42% with EAGLE3, accept 3.6/4.

### What's done

- Replay: capture each verify window's raw projections instead of storing per-position states; recompute the accepted prefix from the committed page on demand.
- Deferred (lazy) commit: accept-time only records a pending window; the actual state write is postponed.
- Fused kernel: the next verify round's kernel replays the pending prefix on its way in — state and conv window stay in registers across the seam (one Triton kernel, constexpr modes: verify / standalone replay / fused, sharing one per-token jit helper so all three are op-identical).
- Lifecycle: pending keyed by request id with page-identity + causal gates; flush on non-verify prep / pause (weight-update fence) / no-forward rounds; departed requests dropped, never flushed. Neutral engine hooks (has/flush/drop_pending on AttentionBackend).
- NVIDIA-only kernel with unchanged AMD behavior (registry-gated; scratch fallback intact). 35 tests incl. 8 adversarial boundary cases.

### CODE REVIEW

**Comment by nperrin-fr** [comment]
kick off ci for now, still wip

File: python/tokenspeed/runtime/layers/attention/backends/hybrid_kda.py [hunk 1]
```diff
@@ -409,6 +568,1046 @@ def _prefill_scan(
...
+        if pending is None:
+            return
+        departed = [
+            r
+            for r, rid in pending["id_by_rpi"].items()
+            if rid is not None and rid not in resident_request_ids
+        ]
+        if departed:
+            self._drop_kda_pending(departed)
+        self._flush_kda_pending()
```
[the remark this answers is not in the document]
Comment (nperrin-fr) [reply python/tokenspeed/runtime/layers/attention/backends/hybrid_kda.py:1231]: Confirmed real, and independently found by our adversarial review (the docstring's "Known gap"). The forward paths screen this correctly (departure/extend-row drops); only the pause fence flushes on residency alone. A correct fix needs the scheduler to expose its Decoding/Retracted set so the fence can drop retracted owners instead of flushing them — dropping unconditionally would lose accepted windows for live requests. Deliberately deferred to a follow-up PR; tracked.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py [hunk 2]
```diff
@@ -1653,32 +1682,172 @@ def try_kda_fused_paged_verify(
...
+        commit_indices=commit_indices,
+        gate_scratch=gate_scratch,
+    )
+    if out is not None and prev_qkv is not None:
+        # A replay round without capture destinations would silently skip the
+        # conv commit and lose this round's payload; say so instead.
+        assert capture is not None, "a replay verify round must supply capture"
+        from tokenspeed_kernel.thirdparty.triton.fla_kda_recurrent import (
+            kda_commit_conv_window,
+        )
```
[the remark this answers is not in the document]
Comment (nperrin-fr) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py]: Fixed in 57b9b7b8: the conv-window commit moved into the registered Triton implementation with capture passed through the dispatch. The wrapper is dispatch-only now, and the impl docstring states the contract: any registered solution that replays a pending window owns its commit.

File: python/tokenspeed/runtime/engine/event_loop.py [hunk 3]
```diff
@@ -1605,6 +1643,8 @@ def event_loop(self):
             self._drain_ready_epd_embeddings()
             self._commit_cache_results()
             if self._pause.forward_blocked:
+                # Retire deferred backend windows while the loop is idle.
+                self._flush_deferred_state()
```
[the remark this answers is not in the document]
Comment (nperrin-fr) [reply python/tokenspeed/runtime/engine/event_loop.py:1641]: Fixed in 57b9b7b8: the handler settles deferred backend state through an engine-wired callback (same pattern as clear_cache_fn) before update_weights_from_distributed loads anything, closing the same-drain window. Unit tests pin the settle-before-load ordering and the no-hook path.

File: python/tokenspeed/runtime/layers/attention/backends/hybrid_kda.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nperrin-fr) [reply python/tokenspeed/runtime/layers/attention/backends/hybrid_kda.py:1231]: Fixed in f2157baa, without a scheduler change: the plan's pages_to_zero already names every freshly re-admitted page per cache group, and admission is the only way a page changes hands — the engine hands that list to the backend before zeroing, and the backend condemns any pending row targeting a listed page via the device slot table (the compose's own idiom, so flush and fuse self-mask). Hand-off happens inside plan building and a fence only fires at iteration start, so condemnation always precedes any fence. Regression test covers retracted-but-resident + recycled pages, and fails with the hook removed.

**Comment by nperrin-fr** [comment]
Both review findings fixed in 9ffff8ca: the replay-support probe now uses the configured dtype (fp32 correctly declines; kernel-side test added), and the condemned page IDs stage through pinned memory with a non-blocking copy so the hook stays sync-free. Validated: runtime suites 67 passed, kernel replay-commit suite 26 passed.

**Comment by nperrin-fr** [comment]
@OftenDream

**Comment by nperrin-fr** [comment]
Fixed in 19320a84. The method sat below the `__main__` guard, so pytest never defined it and the drop-before-zero contract had no coverage. Moved into `ZeroCachePagesContractTest` and verified by negative control: removing the drop call from `zero_cache_pages` fails the test; with it restored the suites pass (46 passed).

**Comment by syuoni** [comment]
# Review: PR #1058 

## Main question: is the flush machinery worth its complexity?

Deferring the commit turns the state write from a stored *result* into a stored
*recipe* (payload + anchor page + accept length + live layer weights), which
stays valid only until the world moves. Hence the escape hatch —
`flush_deferred_state` / `_flush_kda_pending` — with the following trigger
inventory:

| Trigger | Why it flushes | Character |
|---|---|---|
| Owner in a non-verify batch | forward is about to read state pages missing the accepted window | forced — pure correctness |
| Payload ring rebuild | replay's input buffers about to be reallocated | forced |
| Memory release / teardown | weights unmapped, pages freed → replay is an illegal access | forced |
| Identity-less batch | can't match windows to owners | conservative fallback |
| `draft_token_num` change | fused kernel bakes T as constexpr | implementation shortcut |
| Weight update | keeps committed state a single model's computation | invariant hygiene — under keep-mode's off-policy tolerance, is replaying through W_new actually worse than the accepted seam? |

Questions for discussion:

- **Every trigger exists because the commit is deferred**, and the flush *is*
  the eager commit executed late. So in effect the design is eager whenever
  anything is unusual and lazy only on the uninterrupted happy path. Given the
  fusion's measured benefit is kernel-local (~18%) and e2e-neutral, is the
  lazy half worth the lifecycle surface it requires? Is there a workload
  (deeper spec trees, more KDA layers, higher bs) where the fusion shows up
  end-to-end?
- **The lifecycle surface is where the known holes live.** Both issues below
  sit in the drop-vs-flush ownership reasoning, and share a root cause:
  "absent from the current batch / present in the residency set" is used as a
  proxy for request lifecycle, which the scheduler doesn't actually expose.
  Would it make sense to gate the fused path on a real lifecycle signal from
  the scheduler (Decoding / retracted / finished), or alternatively to land
  the replay with an eager commit first and add the fusion as a follow-up once
  that signal exists? The eager variant (standalone `STORE_FINAL` kernel at
  accept time) keeps the full memory win and needs none of the pending-window
  machinery.

## Correctness findings

**[CONFIRMED — silent output corruption] Absent-owner drop loses a live
request's window** (`hybrid_kda.py:843`). `_pending_survives_staging` drops
(never commits) the window of any owner absent from the current verify batch,
assuming "absent ⇒ finished or retracted." The scheduler's capacity-blocker
path violates this: a page-reservation failure in `scheduleDecode` skips a
still-Decoding request from the plan with its pages intact. Its window is
classified as departed and discarded; when it reschedules — with no re-prefill
— it decodes forever from a state missing the accepted tokens. No error is
raised.

**[Author-acknowledged known gap] Retraction + pause commits onto a reassigned
page** (`flush_deferred_state` docstring, `hybrid_kda.py:1215`). A
capacity-retracted owner stays in the engine residency set while its pages are
already freed; a pause landing in that window flushes onto a page that may
belong to another request. Same root cause as the confirmed bug, opposite
polarity (spurious write vs. lost write).

## Secondary findings (verified cleanups, independent of the design question)

1. `KdaAttnBackend.init_cuda_graph_state` (`hybrid_kda.py:1532`) —
   byte-for-byte 39-line duplicate of the base class; delete and inherit.
2. `commit_verified_state` tail / `_ensure_verify_scratch` /
   `preallocate_verify_workspace` (~`hybrid_kda.py:1400`) — re-implement base
   methods with an `if self._replay_active` gate; already drifting from the
   originals. Gate then `super()`.
3. `kda_capture_payload_kernel` (`fla_kda_recurrent.py:1986`, ~90 lines) —
   zero production callers; test oracle only. Move to the test module.
4. `kda_replay_commit_supported` (`ops/attention/__init__.py:1805`) — verbatim
   clone of the GDN probe; extract a shared helper.
5. `_kda_owner_commit_gate` (`hybrid_kda.py:996`) — O(bs²) `list.index` scans;
   build a position dict once.

**Comment by nperrin-fr** [comment]
> # Review: PR #1058
> ## Main question: is the flush machinery worth its complexity?
> Deferring the commit turns the state write from a stored _result_ into a stored _recipe_ (payload + anchor page + accept length + live layer weights), which stays valid only until the world moves. Hence the escape hatch — `flush_deferred_state` / `_flush_kda_pending` — with the following trigger inventory:
> 
> Trigger	Why it flushes	Character
> Owner in a non-verify batch	forward is about to read state pages missing the accepted window	forced — pure correctness
> Payload ring rebuild	replay's input buffers about to be reallocated	forced
> Memory release / teardown	weights unmapped, pages freed → replay is an illegal access	forced
> Identity-less batch	can't match windows to owners	conservative fallback
> `draft_token_num` change	fused kernel bakes T as constexpr	implementation shortcut
> Weight update	keeps committed state a single model's computation	invariant hygiene — under keep-mode's off-policy tolerance, is replaying through W_new actually worse than the accepted seam?
> Questions for discussion:
> 
> * **Every trigger exists because the commit is deferred**, and the flush _is_
>   the eager commit executed late. So in effect the design is eager whenever
>   anything is unusual and lazy only on the uninterrupted happy path. Given the
>   fusion's measured benefit is kernel-local (~18%) and e2e-neutral, is the
>   lazy half worth the lifecycle surface it requires? Is there a workload
>   (deeper spec trees, more KDA layers, higher bs) where the fusion shows up
>   end-to-end?
> * **The lifecycle surface is where the known holes live.** Both issues below
>   sit in the drop-vs-flush ownership reasoning, and share a root cause:
>   "absent from the current batch / present in the residency set" is used as a
>   proxy for request lifecycle, which the scheduler doesn't actually expose.
>   Would it make sense to gate the fused path on a real lifecycle signal from
>   the scheduler (Decoding / retracted / finished), or alternatively to land
>   the replay with an eager commit first and add the fusion as a follow-up once
>   that signal exists? The eager variant (standalone `STORE_FINAL` kernel at
>   accept time) keeps the full memory win and needs none of the pending-window
>   machinery.
> 
> ## Correctness findings
> **[CONFIRMED — silent output corruption] Absent-owner drop loses a live request's window** (`hybrid_kda.py:843`). `_pending_survives_staging` drops (never commits) the window of any owner absent from the current verify batch, assuming "absent ⇒ finished or retracted." The scheduler's capacity-blocker path violates this: a page-reservation failure in `scheduleDecode` skips a still-Decoding request from the plan with its pages intact. Its window is classified as departed and discarded; when it reschedules — with no re-prefill — it decodes forever from a state missing the accepted tokens. No error is raised.
> 
> **[Author-acknowledged known gap] Retraction + pause commits onto a reassigned page** (`flush_deferred_state` docstring, `hybrid_kda.py:1215`). A capacity-retracted owner stays in the engine residency set while its pages are already freed; a pause landing in that window flushes onto a page that may belong to another request. Same root cause as the confirmed bug, opposite polarity (spurious write vs. lost write).
> 
> ## Secondary findings (verified cleanups, independent of the design question)
> 1. `KdaAttnBackend.init_cuda_graph_state` (`hybrid_kda.py:1532`) —
>    byte-for-byte 39-line duplicate of the base class; delete and inherit.
> 2. `commit_verified_state` tail / `_ensure_verify_scratch` /
>    `preallocate_verify_workspace` (~`hybrid_kda.py:1400`) — re-implement base
>    methods with an `if self._replay_active` gate; already drifting from the
>    originals. Gate then `super()`.
> 3. `kda_capture_payload_kernel` (`fla_kda_recurrent.py:1986`, ~90 lines) —
>    zero production callers; test oracle only. Move to the test module.
> 4. `kda_replay_commit_supported` (`ops/attention/__init__.py:1805`) — verbatim
>    clone of the GDN probe; extract a shared helper.
> 5. `_kda_owner_commit_gate` (`hybrid_kda.py:996`) — O(bs²) `list.index` scans;
>    build a position dict once.

Agreed on the complexity, we can do eager commit first.
